lib/redirect-pool.js

```diff
@@ -0,0 +1,187 @@
+
+    // Cannot use RedirectPool on pipelines
+    if (handler.handler) {
+      throw new InvalidArgumentError('RedirectPool cannot be used with pipeline')
+    }
+
+    if (util.isStream(opts.body)) {
+      throw new InvalidArgumentError('body cannot be a stream when using RedirectPool')
+    }
+
+    const maxRedirections = parseMaxRedirections(opts)
+
+    const redirectingHandler = {
+      onConnect (abort) {
+        handler.onConnect(abort)
+      },
+
+      onUpgrade (statusCode, headers, socket) {
+        handler.onUpgrade(statusCode, headers, socket)
+      },
+
+      onError (error) {
+        handler.onError(error)
+      },
+
+      onHeaders (statusCode, headers, resume) {
+        let { [kLeftRedirections]: leftRedirections, [kFollowedRedirections]: followedRedirections } = opts
+
+        if (typeof leftRedirections !== 'number') {
+          leftRedirections = maxRedirections
+        }
+
+        // Check if statusCode is 3xx, if there is a location header and if the redirection can be followed
+        const location = redirectLocation(statusCode, headers, leftRedirections)
+
+        // Nothing to follow, use the original implementation
+        if (!location) {
+          return handler.onHeaders(statusCode, headers, resume, { redirections: followedRedirections })
+        }
+
+        // Gather other request options that will be modified
+        let { method, headers: requestHeaders, body } = opts
+
+        /*
+          https://tools.ietf.org/html/rfc7231#section-6.4
+
+          Remove headers referring to the original URL.
+          By default it is Host only, unless it's a 303 (see below), which removes also all Content-* headers.
+        */
+        if (requestHeaders) {
+          cleanRequestHeaders(requestHeaders, statusCode === 303)
+        }
+
+        // https://tools.ietf.org/html/rfc7231#section-6.4.4
+        // In case of HTTP 303, always replace method to be either HEAD or GET
+        if (statusCode === 303 && method !== 'HEAD') {
+          method = 'GET'
+          body = null
+        }
+
+        // Add the current URL to the list of redirects
+        if (!Array.isArray(followedRedirections)) {
+          followedRedirections = []
+        }
+
+        followedRedirections.push(`${pool[kOrigin]}${opts.path}`)
+
+        // Prepare options for the next request
+        const redirectedOpts = {
+          ...opts,
+          method,
+          path: null,
+          headers: requestHeaders,
+          body,
+          [kLeftRedirections]: leftRedirections - 1,
+          [kFollowedRedirections]: followedRedirections
+        }
+
+        // Follow the redirect using same options - Note the top-level method must be used to support cross origin
+        if (handler.factory) {
+          // This is a StreamHandler
+          stream(location, redirectedOpts, handler.factory, handler.callback)
+        } else {
+          request(location, redirectedOpts, handler.callback)
+        }
```

Comment:
I still think these are not needed. I’ll take a look and try to show within a few days.